### PR TITLE
Token burning

### PR DESCRIPTION
**Pull request by vzotova**
* Burn tokens in `StakingEscrow` (for future reward)
* Ability to burn remaining tokens in `WorkLock`
* Explicit initialization for `StakingEscrow` and `WorkLock`

### CODE REVIEW

File: tests/blockchain/eth/contracts/main/staking_escrow/test_staking_escrow_additional.py [hunk 1]
```diff
@@ -452,7 +453,7 @@ def test_worker(testerchain, token, escrow_contract, deploy_contract):
     worker_log = escrow.events.WorkerSet.createFilter(fromBlock='latest')
 
     # Initialize escrow contract
-    tx = escrow.functions.initialize().transact({'from': creator})
+    tx = escrow.functions.initialize(0).transact({'from': creator})
```
Comment (KPrasch) [inline tests/blockchain/eth/contracts/main/staking_escrow/test_staking_escrow_additional.py]: `0`?

Comment (szotov) [reply tests/blockchain/eth/contracts/main/staking_escrow/test_staking_escrow_additional.py]: Just for tests where mining calculation is not necessary

**Comment by mswilkison** [comment]
The most likely case is that the WorkLock will be uncapped. Since there will be no predetermined exchange rate and the WorkLock allocation will be spread pro rata across all participants, we won't actually use the Burn function.

This could change, so perhaps it makes sense to leave the Burn function in.

**Comment by michwill** [comment]
@mswilkison Well, even in the case of uncapped - we do need burning.
Imagine that someone got # tokens less than minimum needed for staking, and they choose to get their ETH back rather than tokens. What to do with those tokens if not burn?

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol [hunk 2]
```diff
@@ -160,6 +163,16 @@ contract Issuer is Upgradeable {
         currentSupply2 = currentSupply2.sub(_amount);
     }
 
+    /**
+    * @notice Burn sender's tokens. Amount of tokens will be returned for future minting
+    * @param _value Amount to burn
+    **/
+    function burn(uint256 _value) public {
```
Comment (arjunhassard) [inline nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]:
```suggestion
    function burn(uint256 _value) public {
    require(currentMintingPeriod <= 100, "Sorry, you waited too long – get some more Nu and start staking!");
```
If the primary purpose of the burn function is to allow Nu holders who don't meet minimum stake size requirements to get their ETH back, then maybe one should only be able to call it up until a predefined moment in time – e.g. within the first 100 periods of the reward schedule

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol [hunk 3]
```diff
@@ -160,6 +163,16 @@ contract Issuer is Upgradeable {
...
+        token.safeTransferFrom(msg.sender, address(this), _value);
```
Comment (arjunhassard) [inline nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]: Alternatively, or in addition, we could limit the use of this function to participants that end up with less than the min stake size

File: nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol [hunk 4]
```diff
@@ -144,4 +156,16 @@ contract WorkLock {
...
 
+    /**
+    * @notice Burn unclaimed tokens
+    **/
+    function burnRemaining() public {
+        require(block.timestamp >= endBidDate, "Burning tokens allowed when bidding is over");
+        require(remainingTokens > 0, "There are no tokens that can be burned");
+        token.approve(address(escrow), remainingTokens);
+        escrow.burn(remainingTokens);
+        emit Burnt(msg.sender, remainingTokens);
```
Comment (arjunhassard) [inline nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol]: Is there any practical difference in outcome between actively burning your Nu and not claiming them?

Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol]: For now if staker does not claim Nu (in worklock) then can do it latter

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol (hunk 2 above)
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]: Initial idea was for using from worklock and who else wants, but of course we can restrict this method for other Nu holders

File: nucypher/blockchain/eth/sol/source/contracts/Issuer.sol [hunk 5]
```diff
@@ -87,15 +90,15 @@ contract Issuer is Upgradeable {
     /**
     * @notice Initialize reserved tokens for reward
     **/
-    function initialize() public onlyOwner {
+    function initialize(uint256 _reservedReward) public onlyOwner {
         require(currentSupply1 == 0);
+        token.safeTransferFrom(msg.sender, address(this), _reservedReward);
```
[the remark this answers is not in the document]
Comment (szotov) [reply nucypher/blockchain/eth/sol/source/contracts/Issuer.sol]: More control - explicitly set amount of reward, I'm feeling much safer with setting this parameter in function (same for `WorkLock`)

**Pull request closed without merging**